src/table.js

```diff
@@ -542,15 +551,79 @@ export function getTypeValidator(colType) {
   }
 }
 
+export function coerceToType(value, type) {
+  switch (type) {
+    case "string":
+      return typeof value === "string" || value == null
+        ? value
+        : String(value);
+    case "boolean":
+      if (typeof value === "string") {
+        const trimValue = value.trim();
+        return trimValue === "true" ? true : trimValue === "false" ? false : null;
+      }
+      return typeof value === "boolean" || value == null
+        ? value
+        : Boolean(value);
+    case "bigint":
+      return typeof value === "bigint" || value == null
+        ? value
+        : Number.isInteger(typeof value === "string" && !value ? NaN : +value)
+        ? BigInt(value) // eslint-disable-line no-undef
+        : undefined;
+    case "integer": // not a target type for coercion, but can be inferred
+    case "number": {
+      return typeof value === "number"
+        ? value
+        : value == null || (typeof value === "string" && !value)
+        ? NaN
+        : Number(value);
+    }
+    case "date": {
+      if (value instanceof Date) return value;
```

Comment:
I’m strongly tempted to preserve null and undefined here, too.

```suggestion
      if (value instanceof Date || value == null) return value;
```

I understand why we replace null and undefined with NaN for the number type—that’s because we may want to use it in a typed array which doesn’t allow null (and undefined), and we don’t want to null to be coerced to zero because that loses information about missing data. But for date, it’s not clear to me that an invalid Date instance is “better” or “more expected” than null or undefined. What’s principle should we use?

I’m also tempted to add support for coercing numbers to dates; we already support the reverse of coercing dates to numbers implicitly, and it’s really easy:

```js
      if (typeof value === "number") return new Date(value);
```

Reply:
For null and undefined, I agree, I don't see a benefit to returning an invalid Date instance. I think this is what you're saying below, too, @mkfreeman: that if a value is defined and can't be coerced, that's when an invalid Date instance makes sense/may be helpful. But if the value is nullish, I'm not able to think of a reason we would want to return an invalid Date instance for that. Let's use "the principle of least surprise"?? Also agree about adding support for coercing numbers to dates. Thanks for bringing this up, @mbostock.

And @mkfreeman, I initially had the same question: why not pass any defined value to `new Date()`, and let it handle returning valid or invalid Dates. But by using the same regex test we use for inference, we avoid discrepancies between what gets inferred as a date and what gets coerced to a date. (I have a local example of when this can be weird, and would be happy to demo.)